lib/inventory/metadata/metadata.go

```diff
@@ -225,25 +221,45 @@ func (c *fetchConfig) fetchContainerOrchestrator() string {
 
 // fetchCloudEnvironment returns aws, gpc or azure if the instance is running on
 // such cloud environments.
-func (c *fetchConfig) fetchCloudEnvironment() string {
-	if c.awsHTTPGetSuccess() {
-		return "aws"
-	}
-	if c.gcpHTTPGetSuccess() {
-		return "gcp"
-	}
-	if c.azureHTTPGetSuccess() {
-		return "azure"
+func (c *fetchConfig) fetchCloudEnvironment(ctx context.Context) string {
+	ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
+	defer cancel()
+
+	// kick off 3 checks in parallel, at most 1 will succeed
+	cloudEnv := make(chan string, 3)
+
+	go func() {
+		if c.awsHTTPGetSuccess(ctx) {
+			cloudEnv <- "aws"
+		}
+	}()
+
+	go func() {
+		if c.gcpHTTPGetSuccess(ctx) {
+			cloudEnv <- "gcp"
+		}
+	}()
+
+	go func() {
+		if c.azureHTTPGetSuccess(ctx) {
+			cloudEnv <- "azure"
+		}
+	}()
+
+	select {
+	case env := <-cloudEnv:
+		return env
+	case <-ctx.Done():
+		return ""
 	}
```

Comment:
Yeah I'd probably write it like this, I don't think you need to select on the context in this function if the individual checks will reliably return on cancellation
```suggestion
	// kick off checks in parallel, at most 1 will succeed
	checks := []struct{
		env string
		f func(context.Context) bool
	} {
		{ "aws", c.awsHTTPGetSuccess },
		{ "gcp", c.gcpHTTPGetSuccess },
		{ "azure", c.azureHTTPGetSuccess },
	}
	
	cloudEnv := make(chan string, len(checks))
	for _, check := range checks {
		check := check
		go func() {
			if check.f(ctx) {
				cloudEnv <- check.env
			} else {
				cloudEnv <- ""
			}
		}
	}
	
	for _ := range checks {
		if env := <-cloudEnv; env != "" {
			return env
		}
	}
	return ""
```

Reply:
That's a great refactor, thanks!